{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "74549645",
   "metadata": {
    "vscode": {
     "languageId": "latex"
    }
   },
   "source": [
    "Goal: Filtering the dataset to formulate a explicit research question"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "ac7703b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "# load packages\n",
    "import pandas as pd\n",
    "import os\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from tabulate import tabulate # package install: sudo apt install python3-tabulate\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "7f97b5b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 5523 entries, 0 to 5522\n",
      "Data columns (total 30 columns):\n",
      " #   Column            Non-Null Count  Dtype  \n",
      "---  ------            --------------  -----  \n",
      " 0   pdb               5523 non-null   object \n",
      " 1   Hchain            5481 non-null   object \n",
      " 2   Lchain            5344 non-null   object \n",
      " 3   model             5523 non-null   int64  \n",
      " 4   antigen_chain     5490 non-null   object \n",
      " 5   antigen_type      5490 non-null   object \n",
      " 6   antigen_het_name  694 non-null    object \n",
      " 7   antigen_name      5490 non-null   object \n",
      " 8   short_header      5520 non-null   object \n",
      " 9   date              5523 non-null   object \n",
      " 10  compound          5523 non-null   object \n",
      " 11  organism          5522 non-null   object \n",
      " 12  heavy_species     5466 non-null   object \n",
      " 13  light_species     4938 non-null   object \n",
      " 14  antigen_species   5471 non-null   object \n",
      " 15  authors           5523 non-null   object \n",
      " 16  resolution        5523 non-null   float64\n",
      " 17  method            5523 non-null   object \n",
      " 18  r_free            2821 non-null   float64\n",
      " 19  r_factor          2821 non-null   float64\n",
      " 20  scfv              5523 non-null   bool   \n",
      " 21  engineered        5523 non-null   bool   \n",
      " 22  heavy_subclass    5483 non-null   object \n",
      " 23  light_subclass    5344 non-null   object \n",
      " 24  light_ctype       5344 non-null   object \n",
      " 25  affinity          142 non-null    float64\n",
      " 26  delta_g           125 non-null    float64\n",
      " 27  affinity_method   142 non-null    object \n",
      " 28  temperature       80 non-null     object \n",
      " 29  pmid              122 non-null    float64\n",
      "dtypes: bool(2), float64(6), int64(1), object(21)\n",
      "memory usage: 1.2+ MB\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "165690"
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# load csv (comma-separated values) or tsv (tab-separated values)/ other types for even less memory use\n",
    "\n",
    "# good to knows\n",
    "df = pd.read_csv(\"/home/tomstraube/CODING/Data Analysis LINUX/group04-team03/data/ab_ag.tsv\", sep=\"\\t\")\n",
    "df.head() # shows head 5 rows can be index\n",
    "df.tail() # shows tail 5 rows can be index\n",
    "df.columns # shows all column names\n",
    "df.index # How many rows\n",
    "df.index.tolist() # tolist creates List\n",
    "df.info() # useful info columns, Dtype, memory\n",
    "df.describe() # mean, sd\n",
    "df[\"organism\"].unique \n",
    "df.shape \n",
    "df.size\n",
    "#...further -> mirgation to pandas_tutorial...\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d53158cb",
   "metadata": {},
   "source": [
    "First look, what to clean up, what to explore?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04fef7e6",
   "metadata": {},
   "source": [
    "\n",
    "Get rid of empty columns -> Annika\n",
    "Get rid of outliers\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "70b60a3b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Anzahl fehlender Werte pro Spalte:\n",
      "pdb                    0\n",
      "Hchain                42\n",
      "Lchain               179\n",
      "model                  0\n",
      "antigen_chain         33\n",
      "antigen_type          33\n",
      "antigen_het_name    4829\n",
      "antigen_name          33\n",
      "short_header           3\n",
      "date                   0\n",
      "compound               0\n",
      "organism               1\n",
      "heavy_species         57\n",
      "light_species        585\n",
      "antigen_species       52\n",
      "authors                0\n",
      "resolution             0\n",
      "method                 0\n",
      "r_free              2702\n",
      "r_factor            2702\n",
      "scfv                   0\n",
      "engineered             0\n",
      "heavy_subclass        40\n",
      "light_subclass       179\n",
      "light_ctype          179\n",
      "affinity            5381\n",
      "delta_g             5398\n",
      "affinity_method     5381\n",
      "temperature         5443\n",
      "pmid                5401\n",
      "dtype: int64\n",
      "Prozentanzahl NA pro Spalte:\n",
      "pdb                  0.000000\n",
      "Hchain               0.760456\n",
      "Lchain               3.240992\n",
      "model                0.000000\n",
      "antigen_chain        0.597501\n",
      "antigen_type         0.597501\n",
      "antigen_het_name    87.434365\n",
      "antigen_name         0.597501\n",
      "short_header         0.054318\n",
      "date                 0.000000\n",
      "compound             0.000000\n",
      "organism             0.018106\n",
      "heavy_species        1.032048\n",
      "light_species       10.592070\n",
      "antigen_species      0.941517\n",
      "authors              0.000000\n",
      "resolution           0.000000\n",
      "method               0.000000\n",
      "r_free              48.922687\n",
      "r_factor            48.922687\n",
      "scfv                 0.000000\n",
      "engineered           0.000000\n",
      "heavy_subclass       0.724244\n",
      "light_subclass       3.240992\n",
      "light_ctype          3.240992\n",
      "affinity            97.428934\n",
      "delta_g             97.736737\n",
      "affinity_method     97.428934\n",
      "temperature         98.551512\n",
      "pmid                97.791056\n",
      "dtype: float64\n",
      "       pdb Hchain Lchain  model antigen_chain       antigen_type  \\\n",
      "0     9ds2      G      I      0         D | C  protein | protein   \n",
      "1     9ds2      H      J      0         B | A  protein | protein   \n",
      "2     9ds2      K      L      0         F | E  protein | protein   \n",
      "3     8uzp      H      L      0             M            protein   \n",
      "4     8uzp      A      B      0             E            protein   \n",
      "...    ...    ...    ...    ...           ...                ...   \n",
      "5518  4zfg      H      L      0             A            protein   \n",
      "5519  5sy8      H      L      0             O            protein   \n",
      "5520  3idx      H      L      0             G            protein   \n",
      "5521  4ps4      H      L      0             A            protein   \n",
      "5522  7tas      H      L      0             E            protein   \n",
      "\n",
      "                                          antigen_name  \\\n",
      "0     hemagglutinin ha2 chain | hemagglutinin ha1 chai   \n",
      "1     hemagglutinin ha2 chain | hemagglutinin ha1 chai   \n",
      "2     hemagglutinin ha2 chain | hemagglutinin ha1 chai   \n",
      "3                                      stem_mimetic_01   \n",
      "4                                      stem_mimetic_01   \n",
      "...                                                ...   \n",
      "5518                                    angiopoietin-2   \n",
      "5519                      10e8 epitope scaffold t117v2   \n",
      "5520                            hiv-1 hxbc2 gp120 core   \n",
      "5521                                    interleukin-13   \n",
      "5522                                spike glycoprotein   \n",
      "\n",
      "                         short_header      date  \\\n",
      "0         VIRAL PROTEIN/IMMUNE SYSTEM  03/19/25   \n",
      "1         VIRAL PROTEIN/IMMUNE SYSTEM  03/19/25   \n",
      "2         VIRAL PROTEIN/IMMUNE SYSTEM  03/19/25   \n",
      "3                       IMMUNE SYSTEM  03/19/25   \n",
      "4                       IMMUNE SYSTEM  03/19/25   \n",
      "...                               ...       ...   \n",
      "5518  SIGNALING PROTEIN/IMMUNE SYSTEM  04/21/15   \n",
      "5519                    IMMUNE SYSTEM  08/10/16   \n",
      "5520                    IMMUNE SYSTEM  07/22/09   \n",
      "5521                    IMMUNE SYSTEM  03/06/14   \n",
      "5522      VIRAL PROTEIN/IMMUNE SYSTEM  12/21/21   \n",
      "\n",
      "                                               compound  ...  \\\n",
      "0     Crystal structure of 346-54 Fab in complex wit...  ...   \n",
      "1     Crystal structure of 346-54 Fab in complex wit...  ...   \n",
      "2     Crystal structure of 346-54 Fab in complex wit...  ...   \n",
      "3     Crystal Structure of the Computationally Desig...  ...   \n",
      "4     Crystal Structure of the Computationally Desig...  ...   \n",
      "...                                                 ...  ...   \n",
      "5518  Dual-specificity Fab 5A12 in complex with Angi...  ...   \n",
      "5519  Crystal structure of the complex of 10E8 Fab l...  ...   \n",
      "5520  Crystal structure of HIV-gp120 core in complex...  ...   \n",
      "5521  Crystal structure of the complex between IL-13...  ...   \n",
      "5522  SARS-CoV-2 spike in complex with the S2K146 ne...  ...   \n",
      "\n",
      "                                                authors resolution  \\\n",
      "0                               Lin, T.H., Wilson, I.A.      3.290   \n",
      "1                               Lin, T.H., Wilson, I.A.      3.290   \n",
      "2                               Lin, T.H., Wilson, I.A.      3.290   \n",
      "3                           Harshbarger, W., Malito, E.      2.711   \n",
      "4                           Harshbarger, W., Malito, E.      2.711   \n",
      "...                                                 ...        ...   \n",
      "5518                               Harris, S.F., Wu, P.      2.270   \n",
      "5519                           Irimia, A., Wilson, I.A.      1.620   \n",
      "5520  Chen, L., Kwon, Y.D., Zhou, T., Wu, X., O'Dell...      2.500   \n",
      "5521  Teplyakov, A., Obmolova, G., Malia, T., Gillil...      2.800   \n",
      "5522  Park, Y.J., Veesler, D., Seattle Structural Ge...      3.200   \n",
      "\n",
      "                   method r_free r_factor   scfv engineered  heavy_subclass  \\\n",
      "0       X-RAY DIFFRACTION  0.279    0.244  False       True           IGHV3   \n",
      "1       X-RAY DIFFRACTION  0.279    0.244  False       True           IGHV3   \n",
      "2       X-RAY DIFFRACTION  0.279    0.244  False       True           IGHV3   \n",
      "3       X-RAY DIFFRACTION  0.246    0.199  False       True           IGHV1   \n",
      "4       X-RAY DIFFRACTION  0.246    0.199  False       True           IGHV1   \n",
      "...                   ...    ...      ...    ...        ...             ...   \n",
      "5518    X-RAY DIFFRACTION  0.223    0.177  False       True           IGHV3   \n",
      "5519    X-RAY DIFFRACTION  0.214    0.172  False       True           IGHV3   \n",
      "5520    X-RAY DIFFRACTION  0.239    0.178  False       True           IGHV3   \n",
      "5521    X-RAY DIFFRACTION  0.271    0.211  False       True           IGHV2   \n",
      "5522  ELECTRON MICROSCOPY    NaN      NaN  False       True           IGHV3   \n",
      "\n",
      "      light_subclass  light_ctype  \n",
      "0              IGKV3        Kappa  \n",
      "1              IGKV3        Kappa  \n",
      "2              IGKV3        Kappa  \n",
      "3              IGLV1       Lambda  \n",
      "4              IGLV1       Lambda  \n",
      "...              ...          ...  \n",
      "5518           IGKV1        Kappa  \n",
      "5519           IGLV3       Lambda  \n",
      "5520           IGKV1        Kappa  \n",
      "5521          IGKV3D        Kappa  \n",
      "5522           IGLV1       Lambda  \n",
      "\n",
      "[5523 rows x 24 columns]\n"
     ]
    }
   ],
   "source": [
    "# Copy of data_clean.py up\n",
    "# 1. Übersicht über Spalten \n",
    "\n",
    "# Berechnung Anzahl NA pro Spalte \n",
    "na_counts = df.isna().sum()\n",
    "# Ausgabe NA-Anzahl\n",
    "print(\"Anzahl fehlender Werte pro Spalte:\")\n",
    "print(na_counts)\n",
    "\n",
    "#Berechnung Prozent NA pro Spalte \n",
    "na_percent = (na_counts / len(df)) * 100\n",
    "print(\"Prozentanzahl NA pro Spalte:\")\n",
    "print(na_percent)\n",
    "\n",
    "\n",
    "\n",
    "#2. Löschen von NA-Spalten \n",
    "\n",
    "# Umwandlung None in NA \n",
    "\n",
    "df = df.replace({None: np.nan})\n",
    "\n",
    "# Löschung erfolgt nur wenn mind. 85% NA\n",
    "min_non_na_ratio = 0.15\n",
    "\n",
    "# Berechnung absolute Mindestzahl NA\n",
    "min_non_na_count = int(len(df) * min_non_na_ratio)\n",
    "\n",
    "# Spalten über dem Wert löschen \n",
    "df_clean = df.dropna(axis=1, thresh=min_non_na_count)\n",
    "\n",
    "# Speichern als neue .tsv Datei \n",
    "df_clean.to_csv(\"bereinigt.tsv\", sep=\"\\t\", index=False)\n",
    "\n",
    "# Ausgabe in Terminal \n",
    "print(df_clean)\n",
    "\n",
    "\n",
    "#3. Überischt Tabelle df_clean.tsv erstellen\n",
    "\n",
    "# Als Markdown-Tabelle exportieren\n",
    "with open(\"df_clean.tsv\", \"w\") as f:\n",
    "    f.write(tabulate(df_clean, headers='keys', tablefmt='github'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "abaf1414",
   "metadata": {},
   "source": [
    "Analysis of columns for selecting a group of data "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb9d994b",
   "metadata": {},
   "source": [
    "*marked as interesting\n",
    "\n",
    "antigen_name\n",
    "short_header for exp. VIRAL PROTEIN\n",
    "compound for exp. only Cryo-EM structures\n",
    "*organism:\n",
    "    create list of all organism\n",
    "heavy_species and light_species \n",
    "    uninteresting mainly homo sapiens: code for that\n",
    "*antigen species\n",
    "    create list of all species\n",
    "    how many rows has each species and advanced filtering\n",
    "authors\n",
    "    maybe later for deeper analysis\n",
    "*resolution\n",
    "    data with simular resolution \n",
    "    remove rows with useless resolution !Is there data with useless resolution\n",
    "*method\n",
    "    mabye but better focus on specific organism\n",
    "\n",
    "Last need to be understood first\n",
    "\n",
    "r_free \n",
    "r_factor\n",
    "scfv\n",
    "engineerd\n",
    "heavy_subclass\n",
    "light_subclass\n",
    "light_ctype\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "2752b303",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Homo sapiens; Influenza A virus\n",
      "Homo sapiens; Mus musculus\n",
      "Homo sapiens; Severe acute respiratory syndrome coronavirus\n",
      "Homo sapiens\n",
      "Homo sapiens; Severe acute respiratory syndrome coronavirus 2\n",
      "Homo sapiens; SYNTHETIC CONSTRUCT\n",
      "Homo sapiens; Plasmodium falciparum 3D7\n",
      "Homo sapiens; Staphylococcus aureus\n",
      "Homo sapiens; Thermotoga maritima\n",
      "Dendraspis polylepis polylepis; Homo sapiens\n",
      "Homo sapiens; Norovirus Hu/GII.4/Sydney/NSW0514/2012/AU\n",
      "Canis lupus familiaris; Homo sapiens\n",
      "Bos taurus; Homo sapiens\n",
      "dengue virus type 2; Homo sapiens\n",
      "Homo sapiens; Severe acute respiratory syndrome coronavirus 2; synthetic construct; Tequatrovirus T4\n",
      "Homo sapiens; Human immunodeficiency virus 1\n",
      "Homo sapiens; Streptococcus sp. 'group G'\n",
      "Homo sapiens; Influenza A virus (A/California/04/2009(H1N1))\n",
      "Homo sapiens; Rattus norvegicus\n",
      "Homo sapiens; Severe fever with thrombocytopenia syndrome virus\n",
      "Escherichia coli; Homo sapiens\n",
      "Arachis hypogaea; Homo sapiens\n",
      "Homo sapiens; Severe acute respiratory syndrome-related coronavirus\n",
      "Homo sapiens; Zika virus\n",
      "Homo sapiens; Human alphaherpesvirus 3\n",
      "Homo sapiens; Severe acute respiratory syndrome coronavirus 2; Tequatrovirus T4\n",
      "Crimean-Congo hemorrhagic fever virus strain IbAr10200; Homo sapiens\n",
      "Homo sapiens; Human respirovirus 3\n",
      "hepatitis B virus genotype C; Homo sapiens\n",
      "Borreliella burgdorferi B31; Homo sapiens\n",
      "artificial sequences; Homo sapiens\n",
      "Homo sapiens; Plasmodium vivax\n",
      "Homo sapiens; Rabies virus CVS-11\n",
      "Homo sapiens; Respiratory syncytial virus A2\n",
      "Homo sapiens; Plasmodium falciparum HB3\n",
      "Homo sapiens; Metamycoplasma arthritidis\n",
      "Homo sapiens; Plasmodium falciparum\n",
      "Homo sapiens; Lassa virus Josiah\n",
      "Homo sapiens; Influenza A virus (A/Victoria/361/2011(H3N2))\n",
      "Homo sapiens; Human betaherpesvirus 5\n",
      "Henipavirus nipahense; Homo sapiens\n",
      "Echis carinatus; Homo sapiens\n",
      "Homo sapiens; Pseudomonas aeruginosa\n",
      "Hepatitis C virus (isolate 1); Homo sapiens\n",
      "Borreliella burgdorferi; Homo sapiens\n",
      "Enterobacteria phage T6; Homo sapiens; Severe acute respiratory syndrome coronavirus 2; synthetic construct\n",
      "Betapolyomavirus hominis; Homo sapiens\n",
      "Homo sapiens; Saccharopolyspora erythraea\n",
      "Hepacivirus C; Homo sapiens\n",
      "Homo sapiens; Orthohantavirus hantanense\n",
      "Homo sapiens; synthetic construct\n",
      "Crimean-Congo hemorrhagic fever virus; Homo sapiens\n",
      "Homo sapiens; Paslahepevirus balayani\n",
      "Hepatitis E virus rat/R63/DEU/2009; Homo sapiens\n",
      "Clostridium perfringens; Homo sapiens; synthetic construct\n",
      "Homo sapiens; Oplophorus gracilirostris; Rattus norvegicus; SYNTHETIC CONSTRUCT\n",
      "Bos taurus; Homo sapiens; Oplophorus gracilirostris; Rattus norvegicus; SYNTHETIC CONSTRUCT\n",
      "HIV whole-genome vector AA1305#18; Homo sapiens\n",
      "Homo sapiens; Measles morbillivirus\n",
      "Homo sapiens; Influenza A virus (A/Viet Nam/1203/2004(H5N1))\n",
      "Homo sapiens; Influenza A virus (strain A/New Zealand:South Canterbury/35/2000 H1N1)\n",
      "Homo sapiens; Influenza A virus (strain swl A/California/04/2009 H1N1)\n",
      "Enterobacteria phage T4; Homo sapiens; Human respiratory syncytial virus A (strain A2)\n",
      "Epstein-Barr virus (strain GD1); Homo sapiens\n",
      "Homo sapiens; Human alphaherpesvirus 1 strain F\n",
      "Homo sapiens; Human alphaherpesvirus 1\n",
      "Homo sapiens; Human herpesvirus 2 strain G\n",
      "Dermatophagoides pteronyssinus; Homo sapiens\n",
      "Bos taurus; Homo sapiens; Rattus norvegicus\n",
      "Dengue virus type 3; Homo sapiens\n",
      "Deltacoronavirus; Homo sapiens\n",
      "Homo sapiens; Porcine deltacoronavirus\n",
      "Homo sapiens; SFTS virus JS4\n",
      "Homo sapiens; SFTS virus SD4\n",
      "Homo sapiens; SFTS phlebovirus (isolate SFTSV/Human/China/HB29/2010)\n",
      "Homo sapiens; Monkeypox virus\n",
      "Enterobacteria phage T4; Homo sapiens; Severe acute respiratory syndrome coronavirus 2\n",
      "Escherichia coli; Homo sapiens; Oplophorus gracilirostris\n",
      "Homo sapiens; Influenza A virus (A/Shanghai/02/2013(H7N9))\n",
      "Homo sapiens; Human herpesvirus 4\n",
      "Homo sapiens; Xenopus laevis\n",
      "Homo sapiens; Norovirus\n",
      "Homo sapiens; Influenza B virus (B/Iowa/06/2017)\n",
      "Hepacivirus hominis; Homo sapiens\n",
      "Homo sapiens; Ophiophagus hannah\n",
      "Enterobacteria phage T4; Homo sapiens\n",
      "Homo sapiens; Influenza A virus (A/Indiana/10/2011(H3N2))\n",
      "Homo sapiens; Norovirus GI.4\n",
      "Homo sapiens; Human herpesvirus 1 (strain KOS)\n",
      "Homo sapiens; Neisseria meningitidis serogroup B\n",
      "Homo sapiens; Human metapneumovirus\n",
      "Homo sapiens; Respiratory syncytial virus\n",
      "Homo sapiens; Influenza A virus (A/Moscow/10/1999(H3N2))\n",
      "Homo sapiens; Influenza A virus (A/New Caledonia/20/1999(H1N1))\n",
      "Homo sapiens; Rhodopseudomonas palustris\n",
      "Homo sapiens; Influenza A virus (A/Singapore/1/1957(H2N2))\n",
      "Escherichia coli (strain K12); Homo sapiens\n",
      "Bos taurus; Homo sapiens; Influenza A virus (strain A/Victoria/3/1975 H3N2); SYNTHETIC CONSTRUCT\n",
      "Homo sapiens; Sus scrofa\n",
      "Homo sapiens; unidentified influenza virus\n",
      "Homo sapiens; Rift Valley fever virus\n",
      "Homo sapiens; Human respiratory syncytial virus A2\n",
      "Homo sapiens; Human metapneumovirus A\n",
      "Homo sapiens; Plasmodium falciparum (isolate NF54)\n",
      "Homo sapiens; Influenza A virus (A/Brevig Mission/1/1918(H1N1))\n",
      "Homo sapiens; Influenza A virus (A/Sydney/5/1997(H3N2))\n",
      "Escherichia phage T2; Homo sapiens; Human metapneumovirus\n",
      "CENTRUROIDES NOXIUS HOFFMANN; HOMO SAPIENS\n",
      "Hepacivirus C; Homo sapiens; synthetic construct\n",
      "Homo sapiens; Human poliovirus 1 Mahoney\n",
      "Homo sapiens; Human poliovirus 3 strain Sabin\n",
      "Homo sapiens; Poliovirus 2\n",
      "Homo sapiens; Human poliovirus 2 strain Sabin\n",
      "Homo sapiens; Human poliovirus 1 strain Sabin\n",
      "Homo sapiens; Severe acute respiratory syndrome coronavirus 2; SYNTHETIC CONSTRUCT\n",
      "Homo sapiens; Oxyuranus scutellatus scutellatus\n",
      "Homo sapiens; Naja nivea\n",
      "Dendroaspis polylepis polylepis; Homo sapiens\n",
      "Bungarus multicinctus; Homo sapiens\n",
      "Homo sapiens; Lassa mammarenavirus\n",
      "Homo sapiens; Thermotoga maritima MSB8\n",
      "Chikungunya virus strain Senegal 37997; Homo sapiens\n",
      "Homo sapiens; Influenza B virus\n",
      "Homo sapiens; Sin Nombre orthohantavirus\n",
      "Homo sapiens; Pangolin coronavirus\n",
      "Homo sapiens; Tequatrovirus T4\n",
      "Homo sapiens; Homo sapiens neanderthalensis\n",
      "HIV-1 06TG.HT008; Homo sapiens\n",
      "Homo sapiens; Lymphocytic choriomeningitis virus\n",
      "Escherichia coli; Homo sapiens; SYNTHETIC CONSTRUCT\n",
      "Homo sapiens; Human immunodeficiency virus 1; Human immunodeficiency virus type 1\n",
      "Homo sapiens; Human betacoronavirus 2c EMC/2012\n",
      "Homo sapiens; Lama glama\n",
      "Homo sapiens; Streptococcus pyogenes\n",
      "Homo sapiens; Human parainfluenza 3 virus (strain NIH 47885)\n",
      "Homo sapiens; Shewanella oneidensis\n",
      "Homo sapiens; Shewanella oneidensis MR-1\n",
      "Homo sapiens; Middle East respiratory syndrome-related coronavirus\n",
      "Homo sapiens; Zika virus ZIKV/Human/Cambodia/FSS13025/2010\n",
      "Homo sapiens; Zika virus ZIKV/H. sapiens/FrenchPolynesia/10087PF/2013\n",
      "Homo sapiens; Influenza A virus (A/long-tailed duck/Wisconsin/10OS3912/2010(H14N6))\n",
      "Dengue virus 2; Homo sapiens\n",
      "Homo sapiens; Influenza A virus (A/swine/Guangdong/104/2013(H1N1))\n",
      "Homo sapiens; Influenza A virus (A/flat-faced bat/Peru/033/2010(H18N11))\n",
      "Crimean-Congo hemorrhagic fever orthonairovirus; Homo sapiens\n",
      "Homo sapiens; Human papillomavirus type 16; Pongo pygmaeus\n",
      "Homo sapiens; Simian immunodeficiency virus\n",
      "Homo sapiens; Human gammaherpesvirus 4\n",
      "Homo sapiens; Lassa virus\n",
      "Homo sapiens; Yellow fever virus\n",
      "Homo sapiens; Human coronavirus 229E\n",
      "Acetivibrio thermocellus; Homo sapiens\n",
      "Bordetella pertussis; Homo sapiens\n",
      "Homo sapiens; Rotavirus\n",
      "Escherichia coli; Homo sapiens; Trichoplusia ni\n",
      "Homo sapiens; Rabies virus\n",
      "SEVERE ACUTE RESPIRATORY SYNDROME CORONAVIRUS 2; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/TEXAS/50/2012(H3N2)); HOMO SAPIENS\n",
      "HUMAN IMMUNODEFICIENCY VIRUS 1; HOMO SAPIENS\n",
      "HOMO SAPIENS\n",
      "HOMO SAPIENS; BETULA PENDULA\n",
      "HOMO SAPIENS; GALLUS GALLUS\n",
      "INFLUENZA A VIRUS; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/BEIJING/262/1995(H1N1)); HOMO SAPIENS\n",
      "HOMO SAPIENS; SEVERE ACUTE RESPIRATORY SYNDROME CORONAVIRUS 2\n",
      "HOMO SAPIENS; LAMA GLAMA; SYNTHETIC CONSTRUCT\n",
      "HOMO SAPIENS; HUMAN BETAHERPESVIRUS 5\n",
      "HOMO SAPIENS; HUMAN IMMUNODEFICIENCY VIRUS 1\n",
      "BOS TAURUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; PLASMODIUM FALCIPARUM (ISOLATE 3D7)\n",
      "HOMO SAPIENS; INFLUENZA A VIRUS (A/ANHUI/1/2005(H5N1))\n",
      "HOMO SAPIENS; SYNTHETIC CONSTRUCT\n",
      "INFLUENZA A VIRUS (A/ENVIRONMENT/SHANGHAI/S1439/2013(H7N9)); HOMO SAPIENS\n",
      "HUMAN IMMUNODEFICIENCY VIRUS 1; HOMO SAPIENS; SYNTHETIC CONSTRUCT\n",
      "ZIKA VIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN IMMUNODEFICIENCY VIRUS\n",
      "RESPIRATORY SYNCYTIAL VIRUS A2; HOMO SAPIENS\n",
      "PLASMODIUM FALCIPARUM; HOMO SAPIENS\n",
      "INFLUENZA B VIRUS; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/SHANGHAI/JS01/2013(H7N9)); HOMO SAPIENS\n",
      "HOMO SAPIENS; INFLUENZA A VIRUS (A/CHICKEN/VIETNAM/32/2004(H5N1))\n",
      "EBOLA VIRUS SP.; ZAIRE EBOLAVIRUS; HOMO SAPIENS\n",
      "RESPIRATORY SYNCYTIAL VIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; TRITICUM AESTIVUM\n",
      "HOMO SAPIENS; CAMELUS DROMEDARIUS\n",
      "HOMO SAPIENS; MACACA FASCICULARIS\n",
      "PLASMODIUM VIVAX; HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN CORONAVIRUS OC43\n",
      "HOMO SAPIENS; DENGUE VIRUS 2\n",
      "STAPHYLOCOCCUS AUREUS; HOMO SAPIENS\n",
      "EBOLA VIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; CHAETOMIUM THERMOPHILUM (STRAIN DSM 1495 / CBS 144.50 / IMI 039719); CHAETOMIUM THERMOPHILUM\n",
      "HOMO SAPIENS; HEPATITIS C VIRUS (ISOLATE H)\n",
      "HOMO SAPIENS; INFLUENZA A VIRUS\n",
      "PLASMODIUM VIVAX (STRAIN SALVADOR I); HOMO SAPIENS\n",
      "HOMO SAPIENS; ZAIRE EBOLA VIRUS\n",
      "INFLUENZA A VIRUS; INFLUENZA A VIRUS (STRAIN A/HONG KONG/1/1968 H3N2); HOMO SAPIENS\n",
      "HOMO SAPIENS; ENTEROVIRUS A71\n",
      "HOMO SAPIENS; MIDDLE EAST RESPIRATORY SYNDROME CORONAVIRUS\n",
      "HOMO SAPIENS; ENTEROBACTERIA PHAGE T4\n",
      "HUMAN RESPIRATORY SYNCYTIAL VIRUS A (STRAIN A2); HUMAN IMMUNODEFICIENCY VIRUS 1; HOMO SAPIENS\n",
      "HEPACIVIRUS C; HOMO SAPIENS\n",
      "NOROVIRUS HU/GII.4/FARMINGTON HILLS/2004/USA; HOMO SAPIENS\n",
      "DENGUE VIRUS 3; HOMO SAPIENS\n",
      "HOMO SAPIENS; CRIMEAN-CONGO HEMORRHAGIC FEVER VIRUS (STRAIN NIGERIA/IBAR10200/1970)\n",
      "ESCHERICHIA COLI; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/ANHUI/1-BALF_RG44/2013(H7N9)); HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN RESPIRATORY SYNCYTIAL VIRUS\n",
      "ESCHERICHIA COLI O157:H7; HOMO SAPIENS\n",
      "LASSA VIRUS; HOMO SAPIENS; LASSA VIRUS (STRAIN MOUSE/SIERRA LEONE/JOSIAH/1976)\n",
      "MYCOBACTERIUM TUBERCULOSIS H37RV; HOMO SAPIENS\n",
      "ZAIRE EBOLAVIRUS (STRAIN GABON-94); ZAIRE EBOLAVIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; EBOLA VIRUS\n",
      "ARABIDOPSIS THALIANA; HOMO SAPIENS\n",
      "HOMO SAPIENS; ESCHERICHIA COLI\n",
      "HOMO SAPIENS; STREPTOCOCCUS SP. GROUP G\n",
      "HOMO SAPIENS; STAPHYLOCOCCUS AUREUS SUBSP. AUREUS\n",
      "HUMAN RHINOVIRUS 14; HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN SARS CORONAVIRUS\n",
      "HUMAN HERPESVIRUS 3 STRAIN OKA VACCINE; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (STRAIN SWL A/CALIFORNIA/04/2009 H1N1); HOMO SAPIENS\n",
      "SEVERE ACUTE RESPIRATORY SYNDROME CORONAVIRUS 2; TEQUATROVIRUS T4; HOMO SAPIENS\n",
      "HUMAN CORONAVIRUS OC43; HOMO SAPIENS\n",
      "HOMO SAPIENS; ZIKA VIRUS\n",
      "HOMO SAPIENS; HUMAN IMMUNODEFICIENCY VIRUS 1; SYNTHETIC CONSTRUCT\n",
      "HOMO SAPIENS; BACILLUS CIRCULANS\n",
      "HOMO SAPIENS; ZIKA VIRUS ZIKV/H. SAPIENS/FRENCHPOLYNESIA/10087PF/2013\n",
      "DENGUE VIRUS 2; HOMO SAPIENS\n",
      "RATTUS NORVEGICUS; HOMO SAPIENS; ENTEROBACTERIA PHAGE T4\n",
      "HOMO SAPIENS; CIONA INTESTINALIS\n",
      "HOMO SAPIENS; ESCHERICHIA COLI (STRAIN K12)\n",
      "MUS MUSCULUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; PLASMODIUM FALCIPARUM\n",
      "HOMO SAPIENS; STAPHYLOCOCCUS AUREUS (STRAIN BOVINE RF122 / ET3-1)\n",
      "HOMO SAPIENS; METHANOCALDOCOCCUS JANNASCHII (STRAIN ATCC 43067 / DSM 2661 / JAL-1 / JCM 10045 / NBRC 100440)\n",
      "PEPTOCLOSTRIDIUM DIFFICILE; HOMO SAPIENS\n",
      "HOMO SAPIENS; ALKALILIMNICOLA EHRLICHII (STRAIN ATCC BAA-1101 / DSM 17681 / MLHE-1); RUEGERIA POMEROYI (STRAIN ATCC 700808 / DSM 15171 / DSS-3)\n",
      "INFLUENZA A VIRUS (STRAIN A/HONG KONG/1/1968 H3N2); HOMO SAPIENS\n",
      "CLOSTRIDIOIDES DIFFICILE; HOMO SAPIENS\n",
      "HOMO SAPIENS; BOS TAURUS\n",
      "SYNTHETIC CONSTRUCT; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/ANHUI/1/2005(H5N1)); HOMO SAPIENS\n",
      "HUMAN METAPNEUMOVIRUS; HOMO SAPIENS\n",
      "PLASMODIUM VIVAX SAL-1; HOMO SAPIENS\n",
      "HOMO SAPIENS; INFLUENZA A VIRUS (STRAIN A/DUCK/ALBERTA/35/1976 H1N1)\n",
      "HOMO SAPIENS; SACCHAROMYCES CEREVISIAE\n",
      "HOMO SAPIENS; FINEGOLDIA MAGNA\n",
      "DENGUE VIRUS 1; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/MOSCOW/10/1999(H3N2)); HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/SOLOMON ISLANDS/3/2006(H1N1)); HOMO SAPIENS\n",
      "NEISSERIA MENINGITIDIS; HOMO SAPIENS\n",
      "HUMAN BETAHERPESVIRUS 5; HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN ALPHAHERPESVIRUS 3\n",
      "HOMO SAPIENS; LASSA MAMMARENAVIRUS\n",
      "HOMO SAPIENS; DENGUE VIRUS 1\n",
      "SACCHAROMYCES CEREVISIAE; HOMO SAPIENS\n",
      "HENDRA VIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; ARTIFICIAL GENE\n",
      "DENGUE VIRUS 4; HOMO SAPIENS\n",
      "HOMO SAPIENS; FINEGOLDIA MAGNA; STAPHYLOCOCCUS AUREUS; SYNTHETIC CONSTRUCT\n",
      "RECOMBINANT HEPATITIS C VIRUS HK6A/JFH-1; HOMO SAPIENS\n",
      "MIDDLE EAST RESPIRATORY SYNDROME CORONAVIRUS; HOMO SAPIENS\n",
      "HUMAN CYTOMEGALOVIRUS (STRAIN MERLIN); HUMAN CYTOMEGALOVIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; HAEMOPHILUS INFLUENZAE\n",
      "HUMAN RESPIRATORY SYNCYTIAL VIRUS A (STRAIN A2); HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/REASSORTANT/NYMC X-181A(A/NYMC X-157 X CALIFORNIA/07/2009)(H1N1)); HOMO SAPIENS\n",
      "RIFT VALLEY FEVER VIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN CYTOMEGALOVIRUS (STRAIN AD169); HUMAN CYTOMEGALOVIRUS (STRAIN 5508); HUMAN CYTOMEGALOVIRUS; HUMAN CYTOMEGALOVIRUS (STRAIN MERLIN)\n",
      "HOMO SAPIENS; INFLUENZA A VIRUS (A/CANADA/720/2005(H2N2))\n",
      "PLASMODIUM FALCIPARUM (ISOLATE 3D7); HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN RESPIRATORY SYNCYTIAL VIRUS; ENTEROBACTERIA PHAGE T4\n",
      "HOMO SAPIENS; JUNIN MAMMARENAVIRUS\n",
      "INFLUENZA A VIRUS (A/DUCK/FUJIAN/13/2002(H5N1)); HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN RESPIRATORY SYNCYTIAL VIRUS A (STRAIN A2)\n",
      "HOMO SAPIENS; NIPAH HENIPAVIRUS\n",
      "INFLUENZA A VIRUS (A/CALIFORNIA/7/2009(H1N1)); HOMO SAPIENS\n",
      "LASSA VIRUS; LASSA VIRUS (STRAIN MOUSE/SIERRA LEONE/JOSIAH/1976); HOMO SAPIENS\n",
      "HEPATITIS C VIRUS (ISOLATE H); HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (STRAIN A/AICHI/2/1968 H3N2); HOMO SAPIENS\n",
      "HOMO SAPIENS; SACCHAROMYCES CEREVISIAE (STRAIN ATCC 204508 / S288C)\n",
      "HOMO SAPIENS; HIV-1\n",
      "ARTIFICIAL GENE; HOMO SAPIENS\n",
      "YELLOW FEVER VIRUS (STRAIN 17D VACCINE); HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/CALIFORNIA/04/2009(H1N1)); HOMO SAPIENS\n",
      "HUMAN IMMUNODEFICIENCY VIRUS; HOMO SAPIENS; SYNTHETIC CONSTRUCT\n",
      "HUMAN RESPIRATORY SYNCYTIAL VIRUS A2; HUMAN IMMUNODEFICIENCY VIRUS 1; HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN METAPNEUMOVIRUS\n",
      "INFLUENZA A VIRUS (A/CALIFORNIA/04/2009 (H1N1); HOMO SAPIENS\n",
      "HOMO SAPIENS; STAPHYLOCOCCUS AUREUS\n",
      "HOMO SAPIENS; STAPHYLOCOCCUS AUREUS SUBSP. AUREUS TCH60\n",
      "DENGUE VIRUS TYPE 2 (STRAIN THAILAND/NGS-C/1944); HOMO SAPIENS\n",
      "LAKE VICTORIA MARBURGVIRUS (STRAIN RAVN-87); HOMO SAPIENS\n",
      "EBOLA VIRUS; HOMO SAPIENS; ZAIRE EBOLAVIRUS\n",
      "HOMO SAPIENS; HEPACIVIRUS C\n",
      "BAT CORONAVIRUS RATG13; HOMO SAPIENS\n",
      "RECOMBINANT HEPATITIS C VIRUS HK6A/JFH-1; HOMO SAPIENS; STREPTOCOCCUS SP.\n",
      "SEVERE ACUTE RESPIRATORY SYNDROME CORONAVIRUS 2; HUMAN IMMUNODEFICIENCY VIRUS 1; HOMO SAPIENS\n",
      "RABIES LYSSAVIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; DERMATOPHAGOIDES PTERONYSSINUS\n",
      "HOMO SAPIENS; TICK-BORNE ENCEPHALITIS VIRUS (WESTERN SUBTYPE)\n",
      "SEVERE ACUTE RESPIRATORY SYNDROME CORONAVIRUS; HOMO SAPIENS\n",
      "VENEZUELAN EQUINE ENCEPHALITIS VIRUS (STRAIN TC-83); HOMO SAPIENS\n",
      "SEVERE ACUTE RESPIRATORY SYNDROME CORONAVIRUS 2; ENTEROBACTERIA PHAGE T4; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/NEW YORK/107/2003(H7N2)); INFLUENZA A VIRUS; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (STRAIN A/DUCK/CZECHOSLOVAKIA/1956 H4N6); HOMO SAPIENS\n",
      "INFLUENZA A VIRUS; INFLUENZA A VIRUS (A/DENVER/1957(H1N1)); HOMO SAPIENS\n",
      "HOMO SAPIENS; STAPHYLOCOCCUS AUREUS; FINEGOLDIA MAGNA\n",
      "MUS MUSCULUS, HOMO SAPIENS; HOMO SAPIENS\n",
      "HOMO SAPIENS; NEISSERIA MENINGITIDIS\n",
      "VACCINIA VIRUS; HOMO SAPIENS\n",
      "CLOSTRIDIUM TETANI; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/HUNAN/02650/2016(H7N9); HOMO SAPIENS\n",
      "nan\n",
      "LASSA MAMMARENAVIRUS; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/SHANGHAI/02/2013(H7N9)); HOMO SAPIENS\n",
      "SACCHAROPOLYSPORA ERYTHRAEA; HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN IMMUNODEFICIENCY VIRUS; ESCHERICHIA COLI\n",
      "SHEWANELLA ONEIDENSIS; HOMO SAPIENS\n",
      "BUNDIBUGYO EBOLAVIRUS; HOMO SAPIENS\n",
      "HUMAN METAPNEUMOVIRUS; HUMAN IMMUNODEFICIENCY VIRUS 1; HOMO SAPIENS\n",
      "HOMO SAPIENS; MIDDLE EAST RESPIRATORY SYNDROME-RELATED CORONAVIRUS\n",
      "HOMO SAPIENS; WEST NILE VIRUS\n",
      "INFLUENZA A VIRUS (A/CALIFORNIA/07/2009(H1N1)); HOMO SAPIENS\n",
      "HOMO SAPIENS; INFLUENZA A VIRUS (A/SOLOMON ISLANDS/3/2006(H1N1))\n",
      "INFLUENZA A VIRUS (A/VIETNAM/1203/2004(H5N1)); HOMO SAPIENS\n",
      "HOMO SAPIENS; STAPHYLOCOCCUS AUREUS (STRAIN MSSA476)\n",
      "CIONA INTESTINALIS; HOMO SAPIENS\n",
      "HOMO SAPIENS; INFLUENZA A VIRUS (A/THAILAND/1(KAN-1)/2004(H5N1))\n",
      "HOMO SAPIENS; RIFT VALLEY FEVER VIRUS\n",
      "SARS CORONAVIRUS WH20; HOMO SAPIENS\n",
      "HOMO SAPIENS; FINEGOLDIA MAGNA; STAPHYLOCOCCUS AUREUS\n",
      "PHLEUM PRATENSE; HOMO SAPIENS\n",
      "MYCOPLASMA GENITALIUM; HOMO SAPIENS\n",
      "CUPRIAVIDUS METALLIDURANS; HOMO SAPIENS\n",
      "HOMO SAPIENS; CLOSTRIDIOIDES DIFFICILE R20291\n",
      "LASSA VIRUS; HOMO SAPIENS\n",
      "VIBRIO CHOLERAE; HOMO SAPIENS\n",
      "HOMO SAPIENS; MACHUPO VIRUS\n",
      "HUMAN IMMUNODEFICIENCY VIRUS TYPE 1 (JRCSF ISOLATE); HOMO SAPIENS\n",
      "HOMO SAPIENS; SEVERE ACUTE RESPIRATORY SYNDROME CORONAVIRUS\n",
      "MIDDLE EAST RESPIRATORY SYNDROME-RELATED CORONAVIRUS; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/HONG KONG/482/97(H5N1)); HOMO SAPIENS\n",
      "PSEUDOMONAS AERUGINOSA; HOMO SAPIENS\n",
      "HENDRA HENIPAVIRUS; HOMO SAPIENS\n",
      "NORWALK VIRUS; HOMO SAPIENS\n",
      "HUMAN CORONAVIRUS EMC; HOMO SAPIENS\n",
      "MARBURG MARBURGVIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; RECOMBINANT HEPATITIS C VIRUS HK6A/JFH-1\n",
      "HOMO SAPIENS; HUMAN CYTOMEGALOVIRUS (STRAIN TOWNE)\n",
      "ROTAVIRUS A; HOMO SAPIENS\n",
      "HOMO SAPIENS; STAPHYLOCOCCUS AUREUS (STRAIN USA300)\n",
      "INFLUENZA A VIRUS (STRAIN A/JAPAN/305/1957 H2N2); HOMO SAPIENS\n",
      "SACCHAROPOLYSPORA ERYTHRAEA; STREPTOMYCES LASALOCIDI; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS; HUMAN IMMUNODEFICIENCY VIRUS 1; HOMO SAPIENS\n",
      "HOMO SAPIENS; ENTEROVIRUS D68\n",
      "HUMAN IMMUNODEFICIENCY VIRUS; HOMO SAPIENS\n",
      "BAT SARS-LIKE CORONAVIRUS WIV1; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/VIET NAM/1203/2004(H5N1)); HOMO SAPIENS\n",
      "ORYCTOLAGUS CUNICULUS; HOMO SAPIENS\n",
      "SEVERE ACUTE RESPIRATORY SYNDROME-RELATED CORONAVIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; INFLUENZA A VIRUS (STRAIN A/HONG KONG/1/1968 H3N2)\n",
      "NOROVIRUS HU/GII.4/SYDNEY/NSW0514/2012/AU; HOMO SAPIENS\n",
      "STAPHYLOCOCCUS AUREUS (STRAIN USA300); HOMO SAPIENS\n",
      "HUMAN IMMUNODEFICIENCY VIRUS TYPE 1; HOMO SAPIENS\n",
      "PLASMODIUM FALCIPARUM (ISOLATE 7G8); HOMO SAPIENS\n",
      "HOMO SAPIENS; SACCHAROMYCES CEREVISIAE S288C\n",
      "HOMO SAPIENS; HUMAN IMMUNODEFICIENCY VIRUS TYPE 1 (ISOLATE YU2)\n",
      "THERMUS AQUATICUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; RATTUS NORVEGICUS\n",
      "HOMO SAPIENS; DENGUE VIRUS TYPE 1 (STRAIN NAURU/WEST PAC/1974)\n",
      "HUMAN RESPIRATORY SYNCYTIAL VIRUS A (STRAIN A2); HUMAN RESPIRATORY SYNCYTIAL VIRUS A; ESCHERICHIA PHAGE T2; HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN HERPESVIRUS 2\n",
      "ENTEROVIRUS D68; HOMO SAPIENS\n",
      "HUMAN RESPIRATORY SYNCYTIAL VIRUS; HOMO SAPIENS\n",
      "HUMAN CYTOMEGALOVIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; MYCOPLASMA GENITALIUM\n",
      "NEISSERIA MENINGITIDIS SEROGROUP B (STRAIN MC58); HOMO SAPIENS\n",
      "HOMO SAPIENS; PHLEUM PRATENSE\n",
      "HOMO SAPIENS; HUMAN CYTOMEGALOVIRUS\n",
      "HOMO SAPIENS; HOMO SAPIENS, MUS MUSCULUS\n",
      "HOMO SAPIENS; INFLUENZA A VIRUS (A/TURKEY/ITALY/214845/2002(H7N3))\n",
      "SARS CORONAVIRUS MA15; HOMO SAPIENS\n",
      "HOMO SAPIENS; STAPHYLOCOCCUS AUREUS; FINEGOLDIA MAGNA; SYNTHETIC CONSTRUCT\n",
      "HOMO SAPIENS; INFLUENZA A VIRUS (A/HONG KONG/1/1968(H3N2)); INFLUENZA A VIRUS\n",
      "HOMO SAPIENS; HOMO HEIDELBERGENSIS\n",
      "HOMO SAPIENS; CENTRUROIDES NOXIUS HOFFMANN\n",
      "HOMO SAPIENS; SEVERE FEVER WITH THROMBOCYTOPENIA VIRUS\n",
      "HOMO SAPIENS; BORRELIA BURGDORFERI (STRAIN ATCC 35210 / B31 / CIP 102532 / DSM 4680)\n",
      "INFLUENZA A VIRUS (A/ANHUI/1-BALF_RG45/2013(H7N9)); HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN RESPIRATORY SYNCYTIAL VIRUS A (STRAIN A2); HUMAN IMMUNODEFICIENCY VIRUS 1; UNIDENTIFIED\n",
      "HOMO SAPIENS; ORECTOLOBUS MACULATUS\n",
      "HUMAN PARVOVIRUS B19; HOMO SAPIENS\n",
      "HOMO SAPIENS; PLASMODIUM VIVAX (STRAIN SALVADOR I)\n",
      "HUMAN CYTOMEGALOVIRUS; HOMO SAPIENS; HUMAN CYTOMEGALOVIRUS (STRAIN 5508); HUMAN CYTOMEGALOVIRUS (STRAIN AD169); HUMAN CYTOMEGALOVIRUS (STRAIN MERLIN)\n",
      "HOMO SAPIENS; CENTRUROIDES NOXIUS\n",
      "CLOSTRIDIUM DIFFICILE; HOMO SAPIENS\n",
      "HOMO SAPIENS; TICK-BORNE ENCEPHALITIS VIRUS FAR EASTERN SUBTYPE\n",
      "INFLUENZA A VIRUS (STRAIN A/BREVIG MISSION/1/1918 H1N1); HOMO SAPIENS\n",
      "MACACA MULATTA; HOMO SAPIENS\n",
      "HOMO SAPIENS; RESTON EBOLAVIRUS\n",
      "HOMO SAPIENS; TICK-BORNE ENCEPHALITIS VIRUS\n",
      "SARS CORONAVIRUS; HOMO SAPIENS\n",
      "HOMO SAPIENS; SARS CORONAVIRUS\n",
      "HOMO SAPIENS; GLYCINE MAX\n",
      "HUMAN IMMUNODEFICIENCY VIRUS, SYNTHETIC CONSTRUCT, THERMOMONOSPORA FUSCA; HOMO SAPIENS\n",
      "HOMO SAPIENS; HUMAN PARECHOVIRUS 3\n",
      "DENGUE VIRUS; HOMO SAPIENS\n",
      "HUMAN RESPIRATORY SYNCYTIAL VIRUS A; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/SWINE/MINNESOTA/A01134337/2010(H3N2)); INFLUENZA A VIRUS; HOMO SAPIENS\n",
      "YELLOW FEVER VIRUS; HOMO SAPIENS\n",
      "INFLUENZA A VIRUS (A/BREVIG MISSION/1/1918(H1N1)); HOMO SAPIENS\n",
      "BAT SARS-LIKE CORONAVIRUS RSSHC014; HOMO SAPIENS\n",
      "ZIKA VIRUS (ISOLATE ZIKV/HUMAN/FRENCH POLYNESIA/10087PF/2013); HOMO SAPIENS\n",
      "There are 413 different combinations/types of organisms\n"
     ]
    }
   ],
   "source": [
    "# Taking a look on organisms\n",
    "\n",
    "# creating a list of all organism\n",
    "organism = df_clean[\"organism\"].tolist()\n",
    "\n",
    "# exluding muliples by iteration\n",
    "organism_nomuliples = []\n",
    "for i in organism:\n",
    "  if i not in organism_nomuliples:\n",
    "    organism_nomuliples.append(i)\n",
    "\n",
    "# * with sep prints each element after one another\n",
    "print(*organism_nomuliples, sep=\"\\n\")\n",
    "\n",
    "# how many different organism types\n",
    "print(f\"There are {len(organism_nomuliples)} different combinations/types of organisms\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e28184a",
   "metadata": {},
   "source": [
    "Next step: Sum of rows corresponding to a organism type for subtype selecting\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "id": "18fde133",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[186, 10, 902, 2426, 869, 63, 24, 4, 25, 2, 1, 1, 8, 9, 17, 152, 2, 1, 7, 2, 26, 17, 6, 28, 5, 4, 4, 15, 5, 14, 1, 2, 2, 4, 1, 1, 127, 1, 3, 17, 3, 1, 2, 2, 1, 4, 5, 9, 8, 1, 24, 2, 4, 1, 8, 2, 1, 4, 1, 3, 6, 2, 1, 1, 3, 6, 6, 2, 1, 3, 3, 4, 2, 2, 2, 1, 13, 4, 6, 1, 3, 5, 6, 5, 1, 21, 3, 2, 3, 3, 26, 8, 3, 7, 2, 10, 2, 2, 1, 6, 1, 3, 4, 2, 8, 1, 3, 3, 2, 1, 2, 1, 1, 1, 4, 4, 1, 2, 1, 7, 1, 12, 10, 1, 7, 1, 1, 45, 3, 5, 1, 12, 2, 1, 1, 10, 4, 5, 2, 3, 2, 2, 1, 1, 1, 2, 1, 4, 11, 10, 1, 3, 1, 2, 1, 5, 614, 9, 292, 3099, 3, 24, 108, 8, 242, 11, 4, 81, 8, 8, 3, 99, 16, 28, 13, 91, 3, 32, 12, 3, 10, 4, 7, 2, 3, 8, 8, 6, 14, 9, 19, 2, 10, 60, 20, 4, 6, 2, 2, 2, 3, 7, 4, 3, 2, 10, 4, 8, 2, 3, 3, 4, 15, 2, 8, 3, 6, 6, 1, 1, 5, 3, 3, 18, 1, 4, 8, 3, 1, 4, 1, 2, 19, 2, 4, 2, 10, 20, 4, 2, 5, 2, 4, 2, 3, 18, 7, 1, 2, 5, 3, 5, 3, 3, 1, 2, 2, 4, 2, 4, 4, 4, 6, 3, 2, 2, 7, 1, 3, 5, 3, 8, 3, 4, 1, 4, 3, 4, 1, 7, 1, 4, 7, 6, 2, 6, 1, 4, 20, 2, 3, 4, 3, 6, 1, 2, 2, 1, 2, 1, 1, 8, 7, 3, 2, 1, 3, 1, 2, 8, 2, 2, 4, 1, 9, 2, 4, 1, 1, 2, 5, 2, 3, 2, 4, 2, 2, 2, 3, 5, 1, 1, 2, 1, 8, 2, 4, 4, 244, 4, 2, 4, 3, 1, 2, 1, 4, 3, 3, 2, 1, 4, 3, 1, 5, 1, 2, 1, 3, 2, 2, 2, 2, 1, 2, 1, 1, 1, 1, 1, 1, 1, 1, 10, 1, 3, 2, 7, 1, 1, 1, 1, 3, 1, 2, 1, 2, 1, 1, 2, 2, 3, 1, 4, 4, 1, 2, 2, 1, 3, 2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n",
      "412\n",
      "412\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<bound method NDFrame.head of                                                   Name   SUM\n",
       "0                      Homo sapiens; Influenza A virus   186\n",
       "1                           Homo sapiens; Mus musculus    10\n",
       "2    Homo sapiens; Severe acute respiratory syndrom...   902\n",
       "3                                         Homo sapiens  2426\n",
       "4    Homo sapiens; Severe acute respiratory syndrom...   869\n",
       "..                                                 ...   ...\n",
       "407  INFLUENZA A VIRUS (A/SWINE/MINNESOTA/A01134337...     1\n",
       "408                   YELLOW FEVER VIRUS; HOMO SAPIENS     1\n",
       "409  INFLUENZA A VIRUS (A/BREVIG MISSION/1/1918(H1N...     1\n",
       "410   BAT SARS-LIKE CORONAVIRUS RSSHC014; HOMO SAPIENS     1\n",
       "411  ZIKA VIRUS (ISOLATE ZIKV/HUMAN/FRENCH POLYNESI...     1\n",
       "\n",
       "[412 rows x 2 columns]>"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# nesessary for pattern\n",
    "import re\n",
    "\n",
    "# creat df from list\n",
    "df_organism_sum = pd.DataFrame(organism_nomuliples, columns=['Name'])\n",
    "\n",
    "# Counting list corresponding to organism_nomuliples (list), to be added to the df later\n",
    "counts = []\n",
    "# Creat counting list corresponding to organism_nomuliples (list), to be added to the df later\n",
    "# iteration over the list of keywords\n",
    "# re.escape to deleted unwanted patterns from the strings (TODO What exactly happens!)\n",
    "# count contains the sum of rows of a subtype in organism_nomuliples (is iterating)\n",
    "# adding each count into the list counts\n",
    "counts = []\n",
    "for keyword in organism_nomuliples:\n",
    "    pattern = re.escape(keyword)\n",
    "    count = df_clean[\"organism\"].astype(str).str.count(pattern).sum()\n",
    "    counts.append(count)\n",
    "\n",
    "print(counts)\n",
    "\n",
    "# adds a column to df for the SUMs\n",
    "# lenght of List counts and lenght of df must be equal\n",
    "print(len(counts))\n",
    "print(len(organism_nomuliples))\n",
    "df_organism_sum[\"SUM\"] = counts\n",
    "\n",
    "# check result\n",
    "df_organism_sum.head"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46308b79",
   "metadata": {},
   "source": [
    "Potential sub datasets"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "772b0b9c",
   "metadata": {},
   "source": [
    " Homosapiens, \n",
    "Homo sapiens -> 2426\n",
    "Influenza A virus -> 186\n",
    "SARS Cov. 1 -> 902\n",
    "SARS Cov. 2 -> 869\n",
    "    -> how to show the whole output!!! \n",
    "    -> coordinate with other groups"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
